src/rate-limit.coffee

```diff
@@ -53,27 +55,39 @@ module.exports = (robot) ->
       else
         roomOrUser = context.response.message.user.name
 
-      # Construct a key to rate limit on. If the response was from a room
-      # then append the room name to the key. Otherwise, append the user name.
-      listenerAndRoom = listenerID + "_" + roomOrUser
+      # Get the key used for storing data in lastExecutedTime and
+      # lastNotifiedTime
+      if context.listener.options.rateLimits &&
+         context.listener.options.rateLimits.keyFunction
+        # Get key by calling a user-defined function
+        robot.logger.debug "Getting key from keyFunction"
+        key = context.listener.options.rateLimits.keyFunction(context)
+      else
+        # Construct a key to rate limit on. If the response was from a room
+        # then append the room name to the key. Otherwise, append the user name.
+        robot.logger.debug "Getting key from listenerID + _ + roomOrUser"
+        key = listenerID + "_" + roomOrUser
+
+      robot.logger.debug "key = \"#{key}\""
+
       # See if command has been executed recently in the same room (or with the same user)
-      if lastExecutedTime.hasOwnProperty(listenerAndRoom) and
-         lastExecutedTime[listenerAndRoom] > Date.now() - minPeriodMs
+      if lastExecutedTime.hasOwnProperty(key) and
+         lastExecutedTime[key] > Date.now() - minPeriodMs
         # Command is being executed too quickly!
-        robot.logger.debug "Rate limiting " + listenerID + " in " + roomOrUser + "; #{minPeriodMs} > #{Date.now() - lastExecutedTime[listenerAndRoom]}"
+        robot.logger.debug "Rate limiting " + listenerID + " in " + roomOrUser + "; #{minPeriodMs} > #{Date.now() - lastExecutedTime[key]}"
```

Comment:
"Rate limiting #{key}; #{minPeriodMs}....."

A little less readable than before, but I think that's okay.